minifi/minifi-nar-bundles/minifi-framework-bundle/minifi-framework/minifi-framework-core/src/main/java/org/apache/nifi/minifi/c2/command/PropertiesPersister.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    public Boolean persistProperties(Map<String, String> propertiesToUpdate) {
+        int propertyCountToUpdate = validateProperties(propertiesToUpdate);
+        if (propertyCountToUpdate == 0) {
+            return false;
+        }
+        Set<String> propertiesToUpdateKeys = new HashSet<>(propertiesToUpdate.keySet());
+
+        Set<String> updatedProperties = new HashSet<>();
+        try (BufferedReader reader = new BufferedReader(new FileReader(bootstrapFile));
+            BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(bootstrapNewFile, false))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                for (String key : propertiesToUpdateKeys) {
+                    String prefix = key + EQUALS_SIGN;
+                    if (line.startsWith(prefix) || line.startsWith(HASHMARK_SIGN + prefix)) {
+                        line = prefix + propertiesToUpdate.get(key);
+                        updatedProperties.add(key);
+                    }
+                }
+                bufferedWriter.write(line + System.lineSeparator());
+            }
+
+            // add new properties which has no values before
+            propertiesToUpdateKeys.removeAll(updatedProperties);
+            for (String key : propertiesToUpdateKeys) {
+                bufferedWriter.write(key + EQUALS_SIGN + propertiesToUpdate.get(key) + System.lineSeparator());
+            }
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return true;
+    }
+
+    private int validateProperties(Map<String, String> propertiesToUpdate) {
+        Set<UpdatableProperty> updatableProperties = (Set<UpdatableProperty>) updatePropertiesPropertyProvider.getProperties().get(AVAILABLE_PROPERTIES);
+        Map<String, UpdatableProperty> updatablePropertyMap = updatableProperties.stream().collect(Collectors.toMap(UpdatableProperty::getPropertyName, Function.identity()));
+        int propertyCountToUpdate = 0;
+        for (Map.Entry<String, String> entry : propertiesToUpdate.entrySet()) {
+            UpdatableProperty updatableProperty = updatablePropertyMap.get(entry.getKey());
+            if (updatableProperty == null) {
+                throw new IllegalArgumentException("You can not update the requested property through C2 protocol");
+            }
+            if (!Objects.equals(updatableProperty.getPropertyValue(), entry.getValue())) {
+                if (!getValidator(updatableProperty.getValidator())
+                    .map(validator -> validator.validate(entry.getKey(), entry.getValue(), validationContext))
+                    .map(ValidationResult::isValid)
+                    .orElse(true)) {
+                    throw new IllegalArgumentException(String.format("Invalid value for %s", entry.getKey()));
```

Comment:
With this fail fast approach there could be multiple retries while everything is sorted out what do you think about collecting all the issues and throwing exception only at the ned with all the relevant information?

Reply:
Thanks for the changes it will be really helpful if it comes to that.